pkg/network/java/hotspot.go

```diff
@@ -0,0 +1,286 @@
+	dst.Close()
+	if err != nil {
+		return "", nil, err
+	}
+	if err := syscall.Chown(h.root+dstPath, uid, gid); err != nil {
+		return "", nil, err
+	}
+
+	return dstPath, func() {
+		os.Remove(h.root + dstPath)
+	}, nil
+}
+
+func (h *Hotspot) connect() (cleanup func(), err error) {
+	addr, err := net.ResolveUnixAddr("unix", h.socketPath)
+	if err != nil {
+		return nil, err
+	}
+	conn, err := net.DialUnix("unix", nil, addr)
+	if err != nil {
+		return nil, err
+	}
+
+	if err := conn.SetDeadline(time.Now().Add(3 * time.Second)); err != nil {
+		conn.Close()
+		return nil, err
+	}
+	h.conn = conn
+	return func() { h.conn.Close() }, nil
+}
+
+func (h *Hotspot) parseResponse(buf []byte) (returnCommand int, returnCode int, response string) {
+	line := 0
+	scanner := bufio.NewScanner(bytes.NewReader(buf))
+	for scanner.Scan() {
+		s := string(scanner.Bytes())
+		switch line {
+		case 0:
+			returnCommand, _ = strconv.Atoi(s)
+		case 1:
+			if strings.HasPrefix(s, "return code: ") {
+				returnCode, _ = strconv.Atoi(s[len("return code: "):])
+			}
+		}
+		response += s + "\n"
+		line++
+	}
+	return returnCommand, returnCode, response
+}
+
+// command: tailingNull is necessary here to flush command
+//	otherwise the JVM is block waiting for more bytes
+//	This is apply only for some command like : 'load agent.so true'
+func (h *Hotspot) command(cmd string, tailingNull bool) error {
+	if _, err := h.conn.Write([]byte{'1', 0}); err != nil { // Protocol version
+		return err
+	}
+
+	for _, c := range strings.Split(cmd, " ") {
+		cmd := append([]byte(c), 0)
+		if _, err := h.conn.Write(cmd); err != nil {
+			return err
+		}
+	}
+	if tailingNull {
+		if _, err := h.conn.Write([]byte{0}); err != nil {
+			return err
+		}
+	}
+
+	buf := make([]byte, 8192)
+	if _, err := h.conn.Read(buf); err != nil {
+		return err
+	}
+	returnCommand, returnCode, responseText := h.parseResponse(buf)
+
+	if returnCommand != 0 {
+		return fmt.Errorf("command '%s' return command %d code %d\n%s\n", cmd, returnCommand, returnCode, responseText)
+	}
+	return nil
+}
+
+// the (short) protocol is following
+//  o create a file .attach_pid%d
+//  o send a SIGQUIT signal
+//  o wait for socket file created by the java process
+func (h *Hotspot) attachJVMProtocol(uid int, gid int) error {
+	attachPIDPath := func(root string) string {
+		return fmt.Sprintf("%s/.attach_pid%d", root, h.nsPid)
+	}
+
+	attachPath := attachPIDPath(h.root + h.cwd)
+	hook, err := os.OpenFile(attachPath, os.O_RDWR|os.O_CREATE|os.O_TRUNC, 0660)
+	if err != nil {
+		return err
+	}
+	hook.Close()
+	// we don't check Chown() return error here as it can failed on some filesystem
+	_ = syscall.Chown(attachPath, uid, gid)
+	hookUID, _, ownerErr := getPathOwner(attachPath)
+	if err != nil || ownerErr != nil || hookUID != uint32(uid) {
+		// we failed to create the .attach_pid file in the process directory
+		// let's try in /tmp
+
+		// Note: some filesystem over write the owner when creating a file
```

Comment:
a file or the file?

Reply:
it is not clear to me, why some filesystems will override the owner, which owner will it be?